Implement Context Matching

tstadel opened the pull request:
**Proposed changes**:
- Implement methods `calculate_context_similarity` and `match_context`

**Status (please check what you already did)**:
- [x] First draft (up for discussions & feedback)
- [x] Final code
- [x] Added tests
- [x] Updated documentation

closes https://github.com/deepset-ai/haystack/issues/2265

tstadel commented:
Tests to be added...

ArzelaAscoIi reviewed (commented):
Awesome :) Thanks for also adding `match_contexts` !

julian-risch reviewed (commented):
Looks very good already. Looking forward to some tests. Maybe it's better to rename `matching.py` to `context_matching.py`. Otherwise it's too generic. For now `utils` is fine but I could imagine the code also under `modeling/evaluation`. Don't forget to add labels to the PR. 😉

tstadel commented on haystack/utils/matching.py:
```diff
@@ -0,0 +1,127 @@
+
+    :param context: The context to match.
+    :param candidate: The candidate to match the context.
+    :param min_words: The minimum number of words context and candidate need to have in order to be scored.
```
`words` should be generally better understandable then `tokens` by most people. Also `min_seq_len` refers to a sequence of "whatever" because it totally depends on the tokenizer what you're dealing with (words, wordpieces, bytes, etc.). Here we're actually dealing with words. So I'd leave it like that.

tstadel commented:
@julian-risch @ArzelaAscoIi 
First version should be complete now:
- test added
- changed `min_words` into `min_length` param, so we work on string length now which should be more generic and produce less edge cases than words
- added `boost_split_overlaps` param to control whether boosting partial split overlaps resulting from preprocessing should be boosted/matched
- added non-parallel (num_processes<=1) versions of `match_context` and `match_contexts` for easier debugging

ArzelaAscoIi reviewed (approved):
Quite complex! Looks good! 🚀

julian-risch commented on haystack/utils/context_matching.py:
```diff
@@ -0,0 +1,213 @@
+        grouped_matches = groupby(group_sorted_matches, key=lambda candidate: candidate.context_id)
+        for context_id, group in grouped_matches:
+            sorted_group = sorted(group, key=lambda candidate: candidate.score, reverse=True)
+            match_list = list((candiate_score.candidate_id, candiate_score.score) for candiate_score in sorted_group)
```
typo in `candiate_score`

julian-risch reviewed (approved):
LGTM! 👍 There is just a small typo to be fixed before merging. Further, let's keep an eye on whether `boost_split_overlaps` increases the number of false positive matches. In that case, we might not want to use by default and set `boost_split_overlaps=False` by default.

tstadel commented:
I agree the `boost_split_overlaps` flag needs further empirical testing. So far it does its job pretty good regarding the tests. So for now, I leave it with default value `True`.
